### PR TITLE
Add readiness & liveness probes

### LINKED ISSUE #1190: Create liveness/readiness endpoints
For kubernetes and container env. we need to implement the standard of liveness and readiness

I suggest /health/ready
                 /health/alive

ready must return 200 if app is ready (db loaded and seeded)
you can also return NOT 200 if you are too loaded, to receive less traffic
and alive should run few VITAL cheks like is mysql work, is rabbit connection ok yes, like 5 sanity tests

@yivo can you suggest 4-5 simple and fast tests we can do ?

### PR DESCRIPTION
**Pull request by tamadamas**

### CODE REVIEW

File: config/routes.rb [hunk 1]
```diff
@@ -70,6 +70,9 @@ def draw(routes_name)
     post '/pusher/auth', to: 'pusher#auth'
   end
 
+  get 'health/liveness_probe', to: 'public/health#liveness_probe'
```
Comment (calj) [inline config/routes.rb]: I would vote for
/health/ready
/health/alive

File: lib/health_checker.rb [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    def all_checks_passed?(checks)
+      checks.all? { |m| send(m) }
+    end
+
+    def check_db
+      Market.exists?
+    end
+
+    def check_redis
+      result = KlineDB.redis.ping
```
Comment (yivo) [inline lib/health_checker.rb]: Use Rails.cache to access redis.

File: lib/health_checker.rb [hunk 3]
```diff
@@ -0,0 +1,46 @@
+# encoding: UTF-8
+# frozen_string_literal: true
+
+module HealthChecker
+  HealthCheckError = Class.new(StandardError)
```
Comment (yivo) [inline lib/health_checker.rb]: 1) All exceptions must be placed in fixed dir like app/exceptions
2) We don't need this exception. Just remove it. Do simple things.

File: lib/health_checker.rb [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+    def check_db
+      Market.exists?
+    end
+
+    def check_redis
+      result = KlineDB.redis.ping
+      result == 'PONG' or raise HealthCheckError, "Redis.ping returned #{result.inspect}"
+    end
+
+    def check_rabbitmq
```
Comment (yivo) [inline lib/health_checker.rb]: @maksim-litvinov Can you write it in 1 line and use automatic close?

File: lib/health_checker.rb [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+    end
+
+    private
+
+    def all_checks_passed?(checks)
+      checks.all? { |m| send(m) }
+    end
+
+    def check_db
+      Market.exists?
```
Comment (yivo) [inline lib/health_checker.rb]: We need to try to connect to database, we may have zero markets in DB. It is OK.

File: lib/health_checker.rb [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+  HealthCheckError = Class.new(StandardError)
+
+  LIVE_CHECKS = %i[check_db check_redis check_rabbitmq].freeze
+  READY_CHECKS = %i[check_db].freeze
+
+  class << self
+    def alive
+      all_checks_passed? LIVE_CHECKS
+    rescue Exception => e
+      report_exception(e)
```
Comment (yivo) [inline lib/health_checker.rb]: Use `report_exception_to_screen`.

File: lib/health_checker.rb [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+      false
+    end
+
+    def ready
+      all_checks_passed? READY_CHECKS
+    rescue Exception => e
+      report_exception(e)
```
Comment (yivo) [inline lib/health_checker.rb]: Use `report_exception_to_screen`. It may hurt Sentry.

File: lib/health_checker.rb [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+# frozen_string_literal: true
+
+module HealthChecker
+  HealthCheckError = Class.new(StandardError)
+
+  LIVE_CHECKS = %i[check_db check_redis check_rabbitmq].freeze
+  READY_CHECKS = %i[check_db].freeze
+
+  class << self
+    def alive
```
Comment (yivo) [inline lib/health_checker.rb]: Ruby convention is to use `?`: `alive?`.

File: lib/health_checker.rb [hunk 9]
```diff
@@ -0,0 +1,46 @@
+# encoding: UTF-8
+# frozen_string_literal: true
+
+module HealthChecker
```
Comment (yivo) [inline lib/health_checker.rb]: Please put to app/services.

File: lib/health_checker.rb [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+  HealthCheckError = Class.new(StandardError)
+
+  LIVE_CHECKS = %i[check_db check_redis check_rabbitmq].freeze
```
Comment (yivo) [inline lib/health_checker.rb]: Incorrect English: Use LIVENESS_CHECKS.

File: lib/health_checker.rb [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+  READY_CHECKS = %i[check_db].freeze
```
Comment (yivo) [inline lib/health_checker.rb]: Use READINESS_CHECKS

File: lib/health_checker.rb [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+
+  class << self
+    def alive
+      all_checks_passed? LIVE_CHECKS
+    rescue Exception => e
```
Comment (yivo) [inline lib/health_checker.rb]: Better rescue StandardException here.

File: lib/health_checker.rb [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+      report_exception(e)
+      false
+    end
+
+    def ready
+      all_checks_passed? READY_CHECKS
+    rescue Exception => e
```
Comment (yivo) [inline lib/health_checker.rb]: Better rescue StandardException here.

File: lib/health_checker.rb [hunk 14]
```diff
@@ -0,0 +1,46 @@
...
+      report_exception(e)
+      false
+    end
+
+    private
+
+    def all_checks_passed?(checks)
```
Comment (yivo) [inline lib/health_checker.rb]: Try to use `check!` for this method.

File: lib/health_checker.rb [hunk 15]
```diff
@@ -0,0 +1,46 @@
...
+      checks.all? { |m| send(m) }
+    end
+
+    def check_db
+      Market.exists?
+    end
+
+    def check_redis
+      result = KlineDB.redis.ping
+      result == 'PONG' or raise HealthCheckError, "Redis.ping returned #{result.inspect}"
```
Comment (yivo) [inline lib/health_checker.rb]: Try to just return `false`.

File: app/controllers/public/health_controller.rb [hunk 16]
```diff
@@ -0,0 +1,20 @@
...
+    end
+
+    def ready
+      render_health_status HealthChecker.ready
+    end
+
+    private
+
+    def render_health_status(health)
+      head(health ? 200 : 503)
```
Comment (yivo) [inline app/controllers/public/health_controller.rb]: Can you omit brackets?

**Review by yivo — CHANGES_REQUESTED** [review]

**Comment by yivo** [comment]
@maksim-litvinov Please, rebase PR.

File: app/services/services/health_checker.rb [hunk 17]
```diff
@@ -0,0 +1,46 @@
...
+      end
+
+      private
+
+      def check!(checks)
+        checks.all? { |m| send(m) }
+      end
+
+      def check_db
+        ActiveRecord::Base.establish_connection
```
Comment (yivo) [inline app/services/services/health_checker.rb]: @maksim-litvinov Can you prove this will not produce lot of zombie connections?

**Review by yivo — CHANGES_REQUESTED** [review]

File: app/services/services/health_checker.rb (hunk 17 above)
Comment (mod) [reply app/services/services/health_checker.rb]: We could do a simple select of something instead @maksim-litvinov

**Comment by tamadamas** [comment]
screenshot from 2018-06-05 11-27-11

Prove of check_db method

**Review by yivo — APPROVED** [review]

**Comment by yivo** [comment]
@maksim-litvinov Actually this can be only proved by checking mysql connections number

**Pull request merged**